### PR TITLE
[Inference API] Add Azure OpenAI completion support

### PR DESCRIPTION
**Pull request by timgrein**
This PR adds Azure OpenAI as provider for the `completion` task type by integrating [Azure OpenAI's chat completions](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#chat-completions).

There's some room for refactorings, which will be addressed after this PR is merged:
- Consolidation of `AzureOpenAiEmbeddingsServiceSettings` and `AzureOpenAiCompletionServiceSettings`
- The consolidation opens up the possibility to infer the path segments for URI construction directly from the settings removing the need of the model subclasses to implement the currently abstract methods like `resource`, `deployment` etc. (marked the place with a `TODO`)

### Testing

Substitute `{{AZURE_OPEN_AI_API_KEY}}` with your own api key.

Creating the inference endpoint for Azure OpenAI
```
PUT http://localhost:3000/_inference/completion/azure_openai_completion

{
    "service": "azureopenai",
    "service_settings": {
        "resource_name":"resource",
        "deployment_id": "gpt-4",
        "api_version": "2024-02-15-preview",
        "api_key": "{{AZURE_OPEN_AI_API_KEY}}"
    }
}
```

Performing completion inference (for example answering a question):
```
POST http://localhost:3000/_inference/completion/azure_openai_completion

{
    "input": "What is Elastic?"
}
```

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/XContentUtils.java [hunk 1]
```diff
@@ -39,7 +39,7 @@ public static void moveToFirstToken(XContentParser parser) throws IOException {
     public static void positionParserAtTokenAfterField(XContentParser parser, String field, String errorMsgTemplate) throws IOException {
         XContentParser.Token token = parser.nextToken();
 
-        while (token != null && token != XContentParser.Token.END_OBJECT) {
+        while (token != null) {
```
Comment (timgrein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/XContentUtils.java:42]: This was needed to correctly parse Azure OpenAI chat completions responses. I've added an explicit test, which would fail otherwise to `XContentUtilsTests`

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.inference.services.ServiceUtils.extractRequiredString;
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.API_VERSION;
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.DEPLOYMENT_ID;
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.RESOURCE_NAME;
+
+public class AzureOpenAiCompletionServiceSettings implements ServiceSettings, AzureOpenAiRateLimitServiceSettings {
```
Comment (timgrein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:34]: Consolidation of `AzureOpenAiCompletionServiceSettings` and `AzureOpenAiEmbeddingsServiceSettings` through a superclass `AzureOpenAiServiceSettings` will be handled in a separate PR after this one merges to keep this PR a bit more contained. This is tracked through an issue in the corresponding epic.

**Comment by timgrein** [comment]
@elasticsearchmachine update branch

**Comment by timgrein** [comment]
@elasticmachine update branch

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        this.apiVersion = apiVersion;
+        this.rateLimitSettings = Objects.requireNonNullElse(rateLimitSettings, DEFAULT_RATE_LIMIT_SETTINGS);
+    }
+
+    public AzureOpenAiCompletionServiceSettings(StreamInput in) throws IOException {
+        resourceName = in.readString();
+        deploymentId = in.readString();
+        apiVersion = in.readString();
+
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ML_INFERENCE_RATE_LIMIT_SETTINGS_ADDED)) {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:89]: I think we can omit this check because these settings are guaranteed to be after the rate limit settings being introduced.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    public RateLimitSettings rateLimitSettings() {
+        return DEFAULT_RATE_LIMIT_SETTINGS;
+    }
+
+    public String apiVersion() {
+        return apiVersion;
+    }
+
+    @Override
+    public DenseVectorFieldMapper.ElementType elementType() {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:118]: How about we not override this. It should default to returning `null`. Completion doesn't really need the `elementType` right?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+        builder.endObject();
+        return builder;
+    }
+
+    private void toXContentFragmentOfExposedFields(XContentBuilder builder, ToXContent.Params params) throws IOException {
+        builder.field(RESOURCE_NAME, resourceName);
+        builder.field(DEPLOYMENT_ID, deploymentId);
+        builder.field(API_VERSION, apiVersion);
+        rateLimitSettings.toXContent(builder, params);
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:141]: I just merged a change for this: https://github.com/elastic/elasticsearch/pull/108261

We made the decision that until the rate limiting logic is complete (merging this PR: https://github.com/elastic/elasticsearch/pull/107706) we'll hide these fields from being returned to the user.

Basically we need to change this method slightly and have it not write the rate limit settings, so move this line to the `toXContent` method. And this class can extend the base class we created to remove some of the boilerplate stuff like `getFilteredXContentObject`.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            toXContentFragmentOfExposedFields(builder, params);
+
+            builder.endObject();
+            return builder;
+        };
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.ML_INFERENCE_AZURE_OPENAI_EMBEDDINGS;
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:158]: We'll want this to be `ML_INFERENCE_AZURE_OPENAI_COMPLETIONS`.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(resourceName);
+        out.writeString(deploymentId);
+        out.writeString(apiVersion);
+
+        if (out.getTransportVersion().onOrAfter(TransportVersions.ML_INFERENCE_RATE_LIMIT_SETTINGS_ADDED)) {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:167]: We can omit the if-check.

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/external/azureopenai/AzureOpenAiAccountTests.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+package org.elasticsearch.xpack.inference.external.azureopenai;
+
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.xpack.inference.services.azureopenai.embeddings.AzureOpenAiEmbeddingsModel;
+import org.elasticsearch.xpack.inference.services.azureopenai.embeddings.AzureOpenAiEmbeddingsModelTests;
+
+import static org.hamcrest.Matchers.is;
+
+public class AzureOpenAiAccountTests extends ESTestCase {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/external/azureopenai/AzureOpenAiAccountTests.java:16]: Thanks for writing these, I mentioned this in your other PR, I think we can remove the class since it's unused now. Unless I missed some place.

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettingsTests.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        var entity = new AzureOpenAiCompletionServiceSettings("resource", "deployment", "2024", null);
+
+        XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON);
+        entity.toXContent(builder, null);
+        String xContentResult = Strings.toString(builder);
+
+        assertThat(xContentResult, is("""
+            {"resource_name":"resource","deployment_id":"deployment","api_version":"2024","rate_limit":{"requests_per_minute":1440}}"""));
+    }
+
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettingsTests.java:65]: After you pull in my PR's changes, could we add a test that ensures the `rate_limit` field is not returned?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.DEPLOYMENT_ID;
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.RESOURCE_NAME;
+
+public class AzureOpenAiCompletionServiceSettings implements ServiceSettings, AzureOpenAiRateLimitServiceSettings {
+
+    public static final String NAME = "azure_openai_completions_service_settings";
+
+    private static final RateLimitSettings DEFAULT_RATE_LIMIT_SETTINGS = new RateLimitSettings(1_440);
+
+    public static AzureOpenAiCompletionServiceSettings fromMap(Map<String, Object> map) {
```
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:40]: Just a heads up, we haven't propagated this change through each task type yet but we found some issues where I forgot to implement backwards compatibility when `fromMap` is invoked for parsing the configuration from persistent storage. We decided that, when that happens we won't validate the configuration to avoid accidental errors when new fields don't exist in old configurations.

This isn't an immediately problem here because all these settings are new. Essentially we'll pass the context into this function and for the persistent configuration case we won't throw validation errors. We can do that in a follow up PR though.

**Review by jonathan-buttner — COMMENTED** [review]
Looking good! I left a few comments.

File: server/src/main/java/org/elasticsearch/TransportVersions.java [hunk 11]
```diff
@@ -193,6 +193,7 @@ static TransportVersion def(int id) {
     public static final TransportVersion NO_GLOBAL_RETENTION_FOR_SYSTEM_DATA_STREAMS = def(8_650_00_0);
     public static final TransportVersion SHUTDOWN_REQUEST_TIMEOUTS_FIX = def(8_651_00_0);
     public static final TransportVersion INDEXING_PRESSURE_REQUEST_REJECTIONS_COUNT = def(8_652_00_0);
+    public static final TransportVersion ML_INFERENCE_AZURE_OPENAI_COMPLETIONS = def(8_653_00_0);
```
[the remark this answers is not in the document]
Comment (timgrein) [reply server/src/main/java/org/elasticsearch/TransportVersions.java:196]: That's a valid point...maybe naming it after the task type is more accurate `completion` (singular)? But also open to name it chat completions. It depends, if we want to go with our task type name or the endpoint name of the provider. WDYT?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiModel.java [hunk 12]
```diff
@@ -50,6 +58,30 @@ protected AzureOpenAiModel(AzureOpenAiModel model, ServiceSettings serviceSettin
 
     public abstract ExecutableAction accept(AzureOpenAiActionVisitor creator, Map<String, Object> taskSettings);
 
+    public URI buildUriString() throws URISyntaxException {
```
[the remark this answers is not in the document]
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiModel.java]: Adjusted with [Make buildUriString final](https://github.com/elastic/elasticsearch/pull/108352/commits/c8759c6fe62542634c44abe01930d8727d3e7b5b)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+import static org.elasticsearch.xpack.inference.services.ServiceUtils.extractRequiredString;
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.API_VERSION;
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.DEPLOYMENT_ID;
+import static org.elasticsearch.xpack.inference.services.azureopenai.AzureOpenAiServiceFields.RESOURCE_NAME;
+
+public class AzureOpenAiCompletionServiceSettings implements ServiceSettings, AzureOpenAiRateLimitServiceSettings {
+
+    public static final String NAME = "azure_openai_completions_service_settings";
+
+    private static final RateLimitSettings DEFAULT_RATE_LIMIT_SETTINGS = new RateLimitSettings(1_440);
```
[the remark this answers is not in the document]
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java]: Good point! Adjusted with [Adjust default rate limit](https://github.com/elastic/elasticsearch/pull/108352/commits/63fee1fb34f7bc976aa36c62bf52840fb00f2313)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java (hunk 3 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java]: [Omit transport version checks for rate limit settings](https://github.com/elastic/elasticsearch/pull/108352/commits/e2260af24237e40d0d9a1a6660ff3a7677bc9881)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java (hunk 6 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java]: [Use correct minimal supported transport version](https://github.com/elastic/elasticsearch/pull/108352/commits/146e0001e5ab65133fe4d917f799d85c86be36d1)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java (hunk 4 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java]: [Remove element type from AzureOpenAiCompletionServiceSettings](https://github.com/elastic/elasticsearch/pull/108352/commits/a80d868bed1c22162525324b060e21c910b556a3)

**Comment by timgrein** [comment]
@elasticmachine update branch

**Comment by timgrein** [comment]
@elasticmachine update branch

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java (hunk 5 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java]: Thanks for the explanation! Adjusted with [Extend FilteredXContentObject](https://github.com/elastic/elasticsearch/pull/108352/commits/14e90626c85ce1a9e9bc354d63c8609735c2c956)

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettingsTests.java (hunk 9 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettingsTests.java:77]: Added the test with [Add testToFilteredXContent test](https://github.com/elastic/elasticsearch/pull/108352/commits/a5d9b4a41f7656dfb33030b15418c0f8308f6a3c)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java (hunk 10 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/completion/AzureOpenAiCompletionServiceSettings.java:58]: Created an issue to track it: https://github.com/elastic/enterprise-search-team/issues/7466

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/azureopenai/AzureOpenAiCompletionAction.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        Objects.requireNonNull(serviceComponents);
+        Objects.requireNonNull(model);
+        this.sender = Objects.requireNonNull(sender);
+        this.requestCreator = new AzureOpenAiCompletionRequestManager(model, serviceComponents.threadPool());
+        this.errorMessage = constructFailedToSendRequestMessage(model.getUri(), "Azure OpenAI completion");
+    }
+
+    @Override
+    public void execute(InferenceInputs inferenceInputs, TimeValue timeout, ActionListener<InferenceServiceResults> listener) {
+        if (inferenceInputs instanceof DocumentsOnlyInput docsOnlyInput) {
```
[the remark this answers is not in the document]
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/azureopenai/AzureOpenAiCompletionAction.java]: Looks way better, adjusted with [Use early returns in AzureOpenAiCompletionAction](https://github.com/elastic/elasticsearch/pull/108352/commits/ac155be9e62d346bbd58fc1b520c66b0e0f44dc7)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiSecretSettings.java [hunk 15]
```diff
@@ -59,14 +63,24 @@ public static AzureOpenAiSecretSettings fromMap(@Nullable Map<String, Object> ma
         return new AzureOpenAiSecretSettings(secureApiToken, secureEntraId);
     }
 
-    public AzureOpenAiSecretSettings {
+    public AzureOpenAiSecretSettings(@Nullable SecureString apiKey, @Nullable SecureString entraId) {
         Objects.requireNonNullElse(apiKey, entraId);
```
Comment (markjhoy) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiSecretSettings.java:67]: FYI - this was set up as require NonNullElse specifically to ensure only either the ApiKey or the Entra ID are supplied -- we need to ensure we cannot have both.

File: server/src/main/java/org/elasticsearch/TransportVersions.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jonathan-buttner) [reply server/src/main/java/org/elasticsearch/TransportVersions.java]: Yeah good point. Doing something like `AzureOpenAiChatCompletion*` keeps it inline with what we already have for the `OpenAiChatCompletion*` classes. I think my vote is `AzureOpenAiChatCompletion*`.

**Review by jonathan-buttner — APPROVED** [review]
I tested locally 🚀

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiSecretSettings.java (hunk 15 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiSecretSettings.java:67]: Yes, this is checked [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiSecretSettings.java#L43-L47)

File: server/src/main/java/org/elasticsearch/TransportVersions.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (timgrein) [reply server/src/main/java/org/elasticsearch/TransportVersions.java]: I'll address the renaming after merge, tracked through https://github.com/elastic/enterprise-search-team/issues/7484
